### ISSUE TITLE
[Phase 3] Detection Ensemble

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/high-dim-anomaly-detection#4]
## Summary

Implement the complete anomaly detection ensemble: Local Outlier Factor (LOF), Statistical Process Control (SPC), and weighted score aggregation. This phase creates the core detection engine.

**Part of:** Meta Issue #1
**Dependencies:** Phase 1 (base detector), Phase 2 (reduced features)

## Objectives

1. Implement Local Outlier Factor detector
2. Implement Statistical Process Control using Mahalanobis distance
3. Create ensemble with weighted score aggregation
4. Normalize and combine detector outputs

## Files to Create

| File | Lines (est.) | Action | Description |
|------|-------------|--------|-------------|
| `src/detection/lof_detector.py` | 80 | Create | Local Outlier Factor wrapper |
| `src/detection/spc_detector.py` | 100 | Create | Mahalanobis distance SPC |
| `src/detection/ensemble.py` | 150 | Create | Weighted ensemble |
| `tests/test_ensemble.py` | 150 | Create | Ensemble tests |

### LOF Detector
- [ ] Implement `LOFDetector` wrapping sklearn LocalOutlierFactor
- [ ] Use `novelty=True` for scoring new samples
- [ ] Configure `n_neighbors`, `contamination` via YAML
- [ ] Invert scores (sklearn: negative = anomaly; we want: positive = anomaly)
- [ ] Handle edge case of fitting on small datasets

### SPC Detector
- [ ] Implement `SPCDetector` using Mahalanobis distance
- [ ] Compute mean and covariance from training data
- [ ] Add regularization to handle singular covariance matrix
- [ ] Score = Mahalanobis distance from center
- [ ] Configure regularization strength via YAML

### Ensemble
- [ ] Implement `AnomalyEnsemble` combining all three detectors
- [ ] Normalize each detector's scores to [0, 1] range
- [ ] Support configurable weights per detector
- [ ] Aggregate via weighted average
- [ ] Expose individual detector scores for debugging

## Code Snippet: LOFDetector

```python
# src/detection/lof_detector.py
"""Local Outlier Factor detector."""

import numpy as np
from sklearn.neighbors import LocalOutlierFactor
from .base_detector import BaseDetector


class LOFDetector(BaseDetector):
    """Density-based anomaly detection using LOF."""

    def __init__(self, config: dict):
        super().__init__(config)
        lof_config = config.get('detection', {}).get('lof', {})
        self.n_neighbors = lof_config.get('n_neighbors', 20)
        self.contamination = lof_config.get('contamination', 0.1)
        self.model = None

    def fit(self, X: np.ndarray) -> 'LOFDetector':
        # novelty=True allows scoring new samples
        self.model = LocalOutlierFactor(
            n_neighbors=min(self.n_neighbors, len(X) - 1),
            contamination=self.contamination,
            novelty=True
        )
        self.model.fit(X)
        self.is_fitted = True
        return self

    def score_samples(self, X: np.ndarray) -> np.ndarray:
        if not self.is_fitted:
            raise RuntimeError("Detector not fitted")
        # sklearn returns negative scores (more negative = more anomalous)
        # We invert to get positive scores (higher = more anomalous)
        return -self.model.score_samples(X)
```

## Code Snippet: SPCDetector

```python
# src/detection/spc_detector.py
"""Statistical Process Control using Mahalanobis distance."""

import numpy as np
from scipy.spatial.distance import mahalanobis
from .base_detector import BaseDetector


class SPCDetector(BaseDetector):
    """Distribution-based anomaly detection using Mahalanobis distance."""

    def __init__(self, config: dict):
        super().__init__(config)
        spc_config = config.get('detection', {}).get('spc', {})
        self.regularization = spc_config.get('regularization', 0.01)
        self.mean = None
        self.cov_inv = None

    def fit(self, X: np.ndarray) -> 'SPCDetector':
        self.mean = np.mean(X, axis=0)
        cov = np.cov(X.T)

        # Regularize covariance for numerical stability
        cov += self.regularization * np.eye(cov.shape[0])

        # Compute inverse
        self.cov_inv = np.linalg.inv(cov)
        self.is_fitted = True
        return self

    def score_samples(self, X: np.ndarray) -> np.ndarray:
        if not self.is_fitted:
            raise RuntimeError("Detector not fitted")

        scores = np.zeros(len(X))
        for i, x in enumerate(X):
            scores[i] = mahalanobis(x, self.mean, self.cov_inv)
        return scores
```

## Code Snippet: AnomalyEnsemble

```python
# src/detection/ensemble.py
"""Anomaly detection ensemble."""

from typing import Dict
import numpy as np
from .base_detector import BaseDetector
from .isolation_forest import IsolationForestDetector
from .lof_detector import LOFDetector
from .spc_detector import SPCDetector


class AnomalyEnsemble(BaseDetector):
    """Weighted ensemble of anomaly detectors."""

    def __init__(self, config: dict):
        super().__init__(config)
        ensemble_config = config.get('detection', {}).get('ensemble', {})

        self.weights = ensemble_config.get('weights', {
            'isolation_forest': 1.0,
            'lof': 1.0,
            'spc': 1.0
        })

        self.detectors: Dict[str, BaseDetector] = {
            'isolation_forest': IsolationForestDetector(config),
            'lof': LOFDetector(config),
            'spc': SPCDetector(config)
        }

    def fit(self, X: np.ndarray) -> 'AnomalyEnsemble':
        for detector in self.detectors.values():
            detector.fit(X)
        self.is_fitted = True
        return self

    def score_samples(self, X: np.ndarray) -> np.ndarray:
        if not self.is_fitted:
            raise RuntimeError("Ensemble not fitted")

        # Get and normalize scores from each detector
        all_scores = {}
        for name, detector in self.detectors.items():
            raw = detector.score_samples(X)
            all_scores[name] = self._normalize(raw)

        # Weighted average
        total_weight = sum(self.weights.values())
        aggregated = np.zeros(len(X))
        for name, scores in all_scores.items():
            aggregated += (self.weights[name] / total_weight) * scores

        return aggregated

    def _normalize(self, scores: np.ndarray) -> np.ndarray:
        """Min-max normalization to [0, 1]."""
        min_s, max_s = scores.min(), scores.max()
        if max_s - min_s < 1e-10:
            return np.zeros_like(scores)
        return (scores - min_s) / (max_s - min_s)

    def get_individual_scores(self, X: np.ndarray) -> Dict[str, np.ndarray]:
        """Get raw scores from each detector (for debugging)."""
        return {
            name: detector.score_samples(X)
            for name, detector in self.detectors.items()
        }
```

## Verification

```python
from sklearn.datasets import make_blobs
import numpy as np
from src.detection.ensemble import AnomalyEnsemble

# Generate data with one cluster
X_normal, _ = make_blobs(n_samples=500, centers=1, random_state=42)

# Create obvious outlier
X_anomaly = np.array([[10, 10]])

config = {
    'detection': {
        'isolation_forest': {'n_estimators': 100},
        'lof': {'n_neighbors': 20},
        'spc': {'regularization': 0.01},
        'ensemble': {'weights': {'isolation_forest': 1.0, 'lof': 1.0, 'spc': 1.0}}
    }
}

ensemble = AnomalyEnsemble(config)
ensemble.fit(X_normal)

normal_scores = ensemble.score_samples(X_normal)
anomaly_scores = ensemble.score_samples(X_anomaly)

print(f"Normal scores: mean={normal_scores.mean():.3f}, max={normal_scores.max():.3f}")
print(f"Anomaly score: {anomaly_scores[0]:.3f}")
assert anomaly_scores[0] > np.percentile(normal_scores, 95), "Anomaly should score higher!"
```

## Definition of Done

- [ ] LOF scores inversely correlate with local density
- [ ] SPC catches distribution deviations (Mahalanobis > threshold)
- [ ] Ensemble score is weighted combination
- [ ] Known outliers consistently score higher than inliers
- [ ] Individual detector scores accessible
- [ ] Unit tests pass

## Learning Notes

**Concepts to understand:**
- Local Outlier Factor: Density-based outlier score
- Mahalanobis distance: Distance accounting for covariance
- Ensemble methods: Combining multiple weak learners
- Score normalization: Making heterogeneous scores comparable
